layouts/shortcodes/flexcolumns.html

```diff
@@ -0,0 +1,34 @@
+{{- if .IsNamedParams -}}
+  {{- $.Scratch.Set "ratio" (.Get "ratio") -}}
+  {{- $.Scratch.Set "mobile" (.Get "flattenInMobileView") -}}
+{{- else -}}
+  {{- $.Scratch.Set "ratio" (.Get 0) -}}
+  {{- $.Scratch.Set "mobile" (.Get 1) -}}
+{{- end -}}
+
+{{- if eq ($.Scratch.Get "mobile") "true" -}}
+  {{- $.Scratch.Set "mobile" "flatten-in-mobile" -}}
+{{- end -}}
+
+{{- if $.Scratch.Get "ratio" -}}
+  {{- $.Scratch.Set "ratio" (split ($.Scratch.Get "ratio") ":") -}}
+{{- end -}}
+
+<div class="book-columns flex flex-wrap {{ $.Scratch.Get "mobile" }}">
+{{- if $.Scratch.Get "ratio" -}}
+  {{- range $i, $val := split .Inner "<--->" -}}
+    {{- $ratioSplit := $.Scratch.Get "ratio" -}}
+    {{- if gt (len $ratioSplit) $i}}
+      <div style="flex-grow:{{ index $ratioSplit $i }};" class="flex-even markdown-inner">
+        {{- $val | $.Page.RenderString -}}
+      </div>
+    {{- end -}}
+  {{- end -}}
+{{- else -}}
+  {{- range split .Inner "<--->" -}}
+  <div class="flex-even markdown-inner">
+    {{- . | $.Page.RenderString -}}
+  </div>
+  {{- end -}}
+{{- end -}}
+</div>
```

Comment:
A couple of (untested) suggestions:

```suggestion
{{-/*
  This shortcode can be used to turn markdown content into columns with a
  flexible ratio instead of being evenly distributed. It is a more powerful
  alternative to the `columns` shortcode.

  It has two optional parameters, `ratio` and `flattenInMobileView`. They can
  be specified by name or position. For example, these two shortcodes have the
  same functionality:

    {{< flexcolumns ratio="3:7" flattenInMobileView="false" >}}

    {{< flexcolumns "3:7" "false" >}}

  Note that to specify `flattenInMobileView` but not `ratio` without using
  named parameters, you **must** specify the ratio as an empty string:

    {{< flexcolumns "" "false" >}}
  
  The `ratio` parameter should be specified as a string of integers separated
  by colons (`:`), such as `1:2:1` or `1:1:3:1`. If `ratio` is not specified
  (or is set to an empty string, `""`), the columns are evenly distributed.
  For more information on flexible columns, see the docs on MDN:
  https://developer.mozilla.org/en-US/docs/Web/CSS/flex-grow
  
  If `flattenInMobileView` is specified as "true", the `flatten-in-mobile`
  class is added to the container for the columns. This causes the columns to
  collapse into a vertical stack when in mobile view to help make them more
  readable on smaller screens. If `flattenInMobileView` is not specified,
  the columns will still display horizontally, even on smaller screens.
*/-}}
{{- $ratio  := "" -}}
{{- $mobile := "" -}}

{{- if .IsNamedParams -}}
  {{- $ratio = (.Get "ratio") -}}
  {{- if eq (.Get "flattenInMobileView") "true" -}}
    {{- $mobile = " flatten-in-mobile" -}}
  {{- end -}}
{{- else -}}
  {{- $ratio = (.Get 0) -}}
  {{- if eq (.Get 1) "true" -}}
    {{- $mobile = " flatten-in-mobile" -}}
  {{- end -}}
{{- end -}}

{{- with $ratio -}}
  {{- $ratio = split . ":" -}}
{{- end -}}

<div class="book-columns flex flex-wrap{{ $mobile }}">
{{- if $ratio -}}
  {{- range $i, $val := split .Inner "<--->" -}}
    {{- if gt (len $ratio) $i}}
      <div style="flex-grow:{{ index $ratio $i }};" class="flex-even markdown-inner">
        {{- $val | $.Page.RenderString -}}
      </div>
    {{- end -}}
  {{- end -}}
{{- else -}}
  {{- range split .Inner "<--->" -}}
  <div class="flex-even markdown-inner">
    {{- . | $.Page.RenderString -}}
  </div>
  {{- end -}}
{{- end -}}
</div>
```

1. I don't _think_ I've done this so far, but I've started adding doc comments to the top of partials and shortcodes. My eventual goal is to add a little parser tool we can use to generate docs for our shortcodes/partials/etc from the first comment in a templated file. I'm also not _sure_ how the template will react with the code examples in the comment; it *should* be okay, but it might catch fire and needs manual testing.
2. While I did raise the point about `scratch` (I had been using global scratch, which is what you used here, but there's also the possibility of creating a local scratch with the [`newscratch` function][newscratch]), I think in this case there's actually some benefit to using local variables and setting them in the parameter-handling blocks:
   - It declares the parameters with their default values (in this case, empty strings) before anything else.
   - It allows us to immediately process the `flattenInMobileView` parameter to the correct value (`flatten-in-mobile` if set to `true`, empty string otherwise).
   - We can reference these multiply-used variables in the actual HTML generation without having to re-retrieve the value from the scratch map several times.
   - The use of the variables is more immediately clear to someone inspecting the template.
3. I slightly modified the class injection so that `$mobile` defaults to an empty string and is set to ` flatten-in-mobile` is set to keep from appending a trailing space to the class definition when the value is not used; this is not a big deal, but I was already poking at the code 😅 
4. I don't think the ratio splitting with the `with` directive is necessarily better than using `if` in this case, but I wanted to include it in case you weren't familiar with how `with` works; it's basically shorthand for "if this value is set to a nonzero value for this type, make it the context variable in the enclosed block" - so the inside becomes `$ratio = split . ":"` instead of `$ratio = split $ratio ":"`. Because `with` clobbers context, we can't use it in the div declaration without storing the page context in a variable; I also think changing it there would make reading the block _more_ confusing than helpful.



[newscratch]: https://gohugo.io/functions/scratch/#the-local-newscratch

Reply:
Tested this, and aside from one change (your big block comment didn't have a space on either side, so it threw an error! Load-bearing comments are my favorite kind of error) it seems to work. As such I've put it in because it's way clearer.